src/utils/config_filters.cpp

```diff
@@ -46,40 +46,72 @@ bool utils::config_filters::unsigned_matches_if_present(const config& filter, co
 	if(filter[attribute].empty()) {
 		return true;
 	}
+	if(cfg[attribute].empty()) {
+		return false;
+	}
 
 	return in_ranges<int>(cfg[attribute].to_int(0), utils::parse_ranges_unsigned(filter[attribute].str()));
 }
 
-bool utils::config_filters::int_matches_if_present(const config& filter, const config& cfg, const std::string& attribute)
+bool utils::config_filters::int_matches_if_present(const config& filter, const config& cfg, const std::string& attribute, bool no_default_value, int def)
 {
 	if(filter[attribute].empty()) {
 		return true;
 	}
+	if(cfg[attribute].empty() && no_default_value) {
+		return false;
+	}
 
-	return in_ranges<int>(cfg[attribute].to_int(0), utils::parse_ranges_int(filter[attribute].str()));
+	return in_ranges<int>(cfg[attribute].to_int(def), utils::parse_ranges_int(filter[attribute].str()));
 }
 
 bool utils::config_filters::int_matches_if_present_or_negative(
-	const config& filter, const config& cfg, const std::string& attribute, const std::string& opposite)
+	const config& filter, const config& cfg, const std::string& attribute, const std::string& opposite, bool no_default_value, int def)
 {
-	if(int_matches_if_present(filter, cfg, attribute)) {
+	if(int_matches_if_present(filter, cfg, attribute, no_default_value , def)) {
 		return true;
 	}
 
 	// don't check for !cfg[opposite].empty(), as this assumes a default value of zero (similarly to
 	// int_matches_if_present)
 	if(cfg[attribute].empty()) {
-		return in_ranges<int>(-cfg[opposite].to_int(0), utils::parse_ranges_int(filter[attribute].str()));
+		if(cfg[opposite].empty() && no_default_value) {
+			return false;
+		}
+		return in_ranges<int>(-cfg[opposite].to_int(def), utils::parse_ranges_int(filter[attribute].str()));
 	}
 
 	return false;
 }
 
-bool utils::config_filters::double_matches_if_present(const config& filter, const config& cfg, const std::string& attribute)
+bool utils::config_filters::double_matches_if_present(const config& filter, const config& cfg, const std::string& attribute, bool no_default_value, double def)
 {
 	if(filter[attribute].empty()) {
 		return true;
 	}
+	if(cfg[attribute].empty() && no_default_value) {
+		return false;
+	}
 
-	return in_ranges<double>(cfg[attribute].to_double(1), utils::parse_ranges_real(filter[attribute].str()));
+	return in_ranges<double>(cfg[attribute].to_double(def), utils::parse_ranges_real(filter[attribute].str()));
+}
+
+bool utils::config_filters::bool_or_empty(const config& filter, const config& cfg, const std::string& attribute)
+{
+	if(filter[attribute].empty()) {
```

Comment:
Should this be `blank()`?

Reply:
i don't know, what diffrence between empty and blank?

Reply:
`blank()` checks if the attribute exists. `empty()` checks if the attribute exists and if it does then it also checks whether its value is an empty string.

Reply:
i changed code for what if (filter[attribute.empty() and !cfg.has_attribute(attribute) then return true

Reply:
I don't know why that would be necessary? The point was that `filter[attribute].empty()` should be `filter[attribute].blank()`.

Reply:
it's done

Reply:
@CelticMinstrel did you mean just here specifically, or throughout the config_filters.cpp file?

Reply:
I'm not sure. I only looked at this one place. If 'empty()` is being used to check if an attribute exists though, then that's wrong. It should either use `blank()` or `has_attribute()`.

Reply:
ialready replace empty() by has_attribute except in unsigned_matches

Reply:
string_matches_if_present and bool_matches_if_present are only function with empty(), should replace wit has_attribute in these functions or not, i think what empty() must be used for bool because empty attribute means what filter check nor true nor false, and for string, i don't know.

Reply:
that string can take an empty value, ok, because we could check the attributes of this type which are empty, but for bool it is better to continue using empty() because I don't know if we should expect that an unspecified filter is false or true by default and to avoid confusion, the check can only be performed if the value is specified.

Reply:
it's done, onlu bool_if_present continue to use empty() instead of bank() or has_attribute()